luigi/s3.py

```diff
@@ -472,12 +482,20 @@ def listdir(self, path, start_time=None, end_time=None):
                     (not start_time and end_time and last_modified_date < end_time) or  # end defined, prior to end
                     (start_time and end_time and start_time < last_modified_date < end_time)  # both defined, between
                ):
-                yield self._add_path_delimiter(path) + item.key[key_path_len:]
+                if with_key:
+                    yield self._add_path_delimiter(path) + item.key[key_path_len:], item
+                else:
+                    yield self._add_path_delimiter(path) + item.key[key_path_len:]
 
     def list(self, path, start_time=None, end_time=None):  # backwards compat
         key_path_len = len(self._add_path_delimiter(path))
-        for item in self.listdir(path, start_time=start_time, end_time=end_time):
-            yield item[key_path_len:]
+        for path in self.listdir(path, start_time=start_time, end_time=end_time):
+            yield path[key_path_len:]
+
+    def list_with_key(self, path, start_time=None, end_time=None):
+        key_path_len = len(self._add_path_delimiter(path))
+        for path, item in self.listdir(path, start_time=start_time, end_time=end_time, with_key=True):
+            yield path[key_path_len:], item
```

Comment:
Yeah... I couldn't think of a more eloquent way to combine them than the following:

```python
def list(self, path, start_time=None, end_time=None, with_key=False):
    key_path_len = len(self._add_path_delimiter(path))
    if with_key:
        for path, item in self.listdir(path, start_time=start_time, end_time=end_time, with_key=with_key):
            yield path[key_path_len:], item
    else:
        for path in self.listdir(path, start_time=start_time, end_time=end_time):
            yield path[key_path_len:]
```

The issue is that `listdir` return a generator of values when `with_key=False` and a generator of tuples when `with_key=True`. So I'm not sure there's a best way to handle these two generator formats with a single loop.

Perhaps, there's a better way to handle `listdir` when `with_key=True`. Possibly only return `boto.s3.key.Key` (such that listdir only yields a single value)? Thoughts?